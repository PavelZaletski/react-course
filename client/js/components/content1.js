import React from 'react';

export default () => (
    <p>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Praesent eu lacinia tellus. Sed et quam risus. Phasellus sit amet </p>
);