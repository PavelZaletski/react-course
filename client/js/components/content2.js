import React from 'react';

export default class Content2 extends React.PureComponent {
    render() {
        return (
            <p>Nulla rhoncus tempus erat nec facilisis. Duis quis purus ligula. Lorem ipsum dolor sit amet</p>
        );
    }
}