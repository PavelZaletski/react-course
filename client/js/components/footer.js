import React from 'react';

export default function() {
    return (
        <footer>Netflixroulette</footer>
    );
}