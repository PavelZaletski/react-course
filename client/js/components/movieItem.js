import React from 'react';

export default function ({ item }) {
    const Genres = item.genres.map(genre => <span key={genre}>{genre}</span>);

    return (
        <div className="movie-item">
            <img className="movie-item__img" src={item.poster_path} />
            <div className="movie-item__description">
                <h4 className="movie-item__title">{item.title}</h4>
                <span className="movie-item__release-date">{new Date(item.release_date).getFullYear()}</span>
            </div>
            <div className="movie-item__genres">{Genres}</div>
        </div>
    );
}