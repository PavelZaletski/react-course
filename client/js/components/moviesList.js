import React from 'react';
import PropTypes from 'prop-types';
import MovieItem from './movieItem';
import MovieNotFound from './movieNotFound';
import { RadioInput } from './radioInput';

export class MoviesList extends React.Component {
    static propTypes = {
        data: PropTypes.array.isRequired
    };

    state = {
        sortValue: 'release'
    };

    sort(arr) {
        if (this.state.sortValue === 'release'){
            return arr.sort((a, b) => new Date(b.release_date) - new Date(a.release_date));
        } else {
            return arr.sort((a, b) => b.vote_average - a.vote_average);
        }
    }

    changeSorting = (value) => {
        this.setState({
            sortValue: value
        });
    }

    render() {
        const data = this.sort(this.props.data);
        const Items = data.map(item => <MovieItem item={item} key={item.id}/>);
        const { sortValue } = this.state; 

        return (
            data.length ?
            <div className="movies-list">
                <div className="movies-list__header">
                    <span>{data.length} movies found</span>
                    <div className="sortby">
                        <span>Sort by</span>
                        <RadioInput value="release" name="sort" text="Release date" onChange={this.changeSorting} checked={sortValue === 'release'}/>
                        <RadioInput value="rating" name="sort" onChange={this.changeSorting} checked={sortValue === 'rating'} />
                    </div>
                </div>
                <div className="movies-list__items">
                    {Items}
                </div>
            </div>
            :
            <MovieNotFound />
        );
    }
}
