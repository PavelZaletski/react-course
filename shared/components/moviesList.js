// @flow

import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import MovieItem from './movieItem';
import RadioInput from './radioInput';
import { sortMovies, fetchMovies } from '../actions/movies-actions';
import { Movie } from '../types/movie';

type Props = {
  movies: Movie[],
  sortMovies: (value: string) => void,
  fetchMovies: ({search: string, searchBy: string}) => void,
  sortBy: string,
  searchBy: string,
  errorMessage: string,
  urlParams: {search: string}
};

export class MoviesListClass extends React.Component<Props> {
    static propTypes = {
      movies: PropTypes.array,
    };

    changeSorting = (value: string) => {
      this.props.sortMovies(value);
    }

    componentWillMount() {
      if (this.props.urlParams) {
        let { search } = this.props.urlParams;
        search = search.replace('-', ' ');

        this.props.fetchMovies({ search, searchBy: this.props.searchBy });
      }
    }

    componentDidUpdate(prevProps: Props) {
      if (this.props.urlParams) {
        const curr = this.props.urlParams;
        const prev = prevProps.urlParams;

        if (prev.search !== curr.search || prevProps.searchBy !== this.props.searchBy) {
          this.props.fetchMovies({ search: curr.search, searchBy: this.props.searchBy });
        }
      }
    }

    render() {
      const { movies, sortBy, errorMessage } = (this.props);
      const Items = movies.map(item => <MovieItem item={item} key={item.id}/>);

      return (
        errorMessage ?
                <div>{errorMessage}</div>
          :
                <div className="movies-list">
                    <div className="movies-list__header">
                        <span>{movies.length} movies found</span>
                        <div className="sortby">
                            <span>Sort by</span>
                            <RadioInput
                              value="release"
                              name="sort"
                              text="Release date"
                              onChange={this.changeSorting}
                              checked={sortBy === 'release'}
                            />
                            <RadioInput
                              value="rating"
                              name="sort"
                              onChange={this.changeSorting}
                              checked={sortBy === 'rating'}
                            />
                        </div>
                    </div>
                    <div className="movies-list__items">
                        {Items}
                    </div>
                </div>
      );
    }
}

const mapStateToProps = store => ({
  sortBy: store.movies.sortBy,
  movies: store.movies.movies,
  errorMessage: store.movies.errorMessage,
  searchBy: store.movies.searchBy,
});

const mapDispatchToProps = dispatch => ({
  sortMovies: (value) => {
    dispatch(sortMovies(value));
  },

  fetchMovies: (params) => {
    dispatch(fetchMovies(params));
  },

});

export const MoviesList = connect(mapStateToProps, mapDispatchToProps)(MoviesListClass);
