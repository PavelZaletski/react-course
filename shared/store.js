import { applyMiddleware, createStore } from 'redux';
import reducer, { rootSaga } from './reducers';
import createSagaMiddleware, { END } from 'redux-saga';

const sagaMiddleware = createSagaMiddleware();

export const getStore = (initialState) => {
  const store = createStore(reducer, initialState, applyMiddleware(sagaMiddleware));

  sagaMiddleware.run(rootSaga);
  store.runSaga = () => sagaMiddleware.run(rootSaga);
  store.close = () => store.dispatch(END);

  return store;
};

export const store = getStore();
