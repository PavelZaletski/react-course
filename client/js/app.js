import React from 'react';
import ReactDOM from 'react-dom';
import { SearchForm } from './components/searchForm';
import { MoviesList } from './components/moviesList';
import { fetchMovies } from './actions/movies-actions';
import { MoviePage } from './components/moviePage';
import { ErrorBoundary } from './components/errorBoundary';
import Footer from './components/footer';

export class App extends React.Component {
    state = {
        movies: [],
        fetched: false
    };

    searchHandler = (params) => {
        fetchMovies(params).then(({data: {data}}) => {
            this.setState({
                movies: data,
                fetched: true
            });
        });
    }

    render() {
        return (
            <ErrorBoundary>
                <SearchForm onSearch={this.searchHandler}/>
                {this.state.fetched && <MoviesList data={this.state.movies} /> }
                {this.state.fetched && <MoviePage movie={this.state.movies[0]} />}
                <Footer />
            </ErrorBoundary>
        );
    }
}
