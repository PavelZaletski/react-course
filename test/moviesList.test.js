import renderer from 'react-test-renderer';
import Enzyme, { shallow, mount } from 'enzyme';
import React from 'react';
import { MoviesList } from '../client/js/components/moviesList';
import { Provider } from 'react-redux';
import { getStore } from '../client/js/store';

describe('MoviesList', () => {
    it('should match snapshot', () => {
        const data = [{ id: 1, genres: [] }, { id: 2, genres: [] }];

        let initialState = {
            movies: {
                movies: data
            }
        };

        const store = getStore(initialState);

        const component = renderer.create(
            <Provider store={store}>
                <MoviesList />
            </Provider>
        ).toJSON();

        expect(component).toMatchSnapshot();
    });
});