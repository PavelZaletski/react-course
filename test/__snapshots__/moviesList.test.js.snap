// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoviesList should match snapshot 1`] = `
<div
  className="movies-list"
>
  <div
    className="movies-list__header"
  >
    <span>
      2
       movies found
    </span>
    <div
      className="sortby"
    >
      <span>
        Sort by
      </span>
      <label
        className=""
      >
        <input
          name="sort"
          onChange={[Function]}
          type="radio"
          value="release"
        />
        <span>
          Release date
        </span>
      </label>
      <label
        className=""
      >
        <input
          name="sort"
          onChange={[Function]}
          type="radio"
          value="rating"
        />
        <span>
          rating
        </span>
      </label>
    </div>
  </div>
  <div
    className="movies-list__items"
  >
    <div
      className="movie-item"
    >
      <img
        className="movie-item__img"
        src={undefined}
      />
      <div
        className="movie-item__description"
      >
        <h4
          className="movie-item__title"
        />
        <span
          className="movie-item__release-date"
        >
          NaN
        </span>
      </div>
      <div
        className="movie-item__genres"
      />
    </div>
    <div
      className="movie-item"
    >
      <img
        className="movie-item__img"
        src={undefined}
      />
      <div
        className="movie-item__description"
      >
        <h4
          className="movie-item__title"
        />
        <span
          className="movie-item__release-date"
        >
          NaN
        </span>
      </div>
      <div
        className="movie-item__genres"
      />
    </div>
  </div>
</div>
`;
